agenta-backend/agenta_backend/middlewares/api_key_manager.py

```diff
@@ -0,0 +1,229 @@
+    - delete_api_key(key): Deletes an API key with the provided key.
+    - is_valid_api_key(key): Checks if an API key is valid by verifying that it is not blacklisted and not expired.
+    - use_api_key(key): Uses an API key and checks rate limiting.
+    """
+
+    def __init__(self):
+        self.redis_client = redis.from_url(url=settings.redis_url)  
+
+    def create_api_key(self, user_id, rate_limit, expiration_date=None):
+        """
+        Creates a new API key with an encrypted user ID, rate limit, and optional expiration date.
+
+        Args:
+        - user_id: The user ID to be encrypted and used as the API key.
+        - rate_limit: The rate limit for the API key.
+        - expiration_date: Optional expiration date for the API key.
+
+        Raises:
+        - ValueError: If a unique key cannot be generated after 5 trials.
+        """
+        for _ in range(5):
+            # Encrypt the user_id using Fernet encryption
+            encrypted_user_id = fernet.encrypt(user_id.encode())
+
+            # Check if an API key with the same key exists
+            existing_key = engine.find_one(APIKeyDB, APIKeyDB.key == encrypted_user_id)
+
+            if not existing_key:
+                # Create an APIKeyDB instance with the encrypted user_id as the key
+                api_key = APIKeyDB(
+                    key=encrypted_user_id,
+                    rate_limit=rate_limit,
+                    expiration_date=expiration_date,
+                )
+
+                # Store the API key in the database
+                engine.save(api_key)
+                return encrypted_user_id
+
+        # If after 5 trials, a unique key is not generated, raise an error
+        raise ValueError("Unable to generate a unique API key")
+
+    def update_api_key(self, key, api_key):
+        """
+        Updates an existing API key with the provided key and APIKeyDB instance.
+
+        Args:
+        - key: The key of the API key to be updated.
+        - api_key: The updated APIKeyDB instance.
+
+        Raises:
+        - KeyError: If the API key does not exist.
+        """
+        # Check if the API key exists
+        existing_key = engine.find_one(APIKeyDB, APIKeyDB.key == key)
+        if not existing_key:
+            raise KeyError("API key not found")
+
+        # Update the API key
+        engine.save(api_key)
+
+    def delete_api_key(self, key):
+        """
+        Deletes an API key with the provided key.
+
+        Args:
+        - key: The key of the API key to be deleted.
+
+        Raises:
+        - KeyError: If the API key does not exist.
+        """
+        # Check if the API key exists
+        existing_key = engine.find_one(APIKeyDB, APIKeyDB.key == key)
+        if not existing_key:
+            raise KeyError("API key not found")
+
+        # Delete the API key
+        engine.delete(existing_key)
+
+    def is_valid_api_key(self, key):
```

Comment:
Nope. Wrong.

We don't care to check whether an api key is valid. We want to extract the user_id from the api key (and return none, which means invalid if not).

So the function would take the api_key provided in the call. It would then hash it (encrypt it), and look for it in the APIKeyDB then return the user_id in that row. If it does not find the encrypted api key, then it would decide that the api key does not exist